[CCP] Implemented separated outcomes tables with select popups

noranda opened the pull request:
**This PR implements the separated outcomes tables with select popups as designed in the [MVP](https://www.figma.com/proto/rLpubNAfrqV99exbLVFz5c/core-cancer-pilot?node-id=555%3A6877&viewport=1100%2C773%2C1&scaling=min-zoom) with the following exceptions to be implemented in future tasks:**
- table/icons toggle
- icons visualization
- side effects - all column
- side effects - header dropdown
- side effects - more (...) button
- legend
- other deaths in bar charts

**This PR does the following:**
- increased max memory to handle large mcode datasets
- changed the ccp patient stage from `IA` to `IIA`
- implemented new structure for outcomes table:
  - added `TreatmentsPopover` component to handle treatment selections
  - added `isSame` and `getCombinations` array utils to help with logic regarding which treatment rows to render
  - updated treatment names object
  - updated `includedTreatments` and `combinedTreatments` default values
  - added event listeners to be able to close poppers by clicking outside of them
  - wrote logic to handle various selections of treatments -- for example:
    - there should be no duplicate rows
    - rows will only show up if there is at least 1 patient with the specified treatment combination
    - if no patients exist to create a row for the included table, helper text will appear -- for the compared row, nothing will appear
- updated `BarChart` component to take in a `comparedRow` and defaulted `active` to false for the time being -- this will be updated in a future task when `BarChart` will be restructured to deal with 'other deaths'
- hooked up similar patient checkboxes for `pathology` and `treatment history`
- styled all components
- fixed and wrote new tests for `TreatmentOptionsOutcome` component
- updated the redux store dev tools extension to get rid of console deprecation warning

****Note: When testing this branch, unselect all similar patient criteria to populate the table, then check off hooked up boxes.**

**All backend and mcode tests pass. Tested in Chrome and Safari.**

_Pull requests into Flux require the following. Submitter should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:. Reviewers should complete every item on the bulleted list when reviewing._


## Submitter:

**Running the Application**

- [x] Manually tested in Chrome (primary browser for testing)
- [x] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [x] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [x] Branch implements task as expected
- [x] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [x] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- [ ] Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- [ ] Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- [ ] Document any new APIs/extension points
- [ ] Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [x] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [x] Code is commented

**Tests**

- [x] Existing tests passed
- [x] Added backend tests for new functionality added


## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

dehall commented on package.json:
```diff
@@ -84,7 +84,7 @@
     "start-js": "react-scripts start",
     "check-version": "node ./scripts/check-version.js",
-    "start": "npm-run-all -p watch-css start-js check-version",
+    "start": "NODE_OPTIONS=--max-old-space-size=5120 npm-run-all -p watch-css start-js check-version",
```
This syntax doesn't work in Windows. One possible fix is to add the `cross-env` library and make the call `cross-env NODE_OPTIONS=--max-old-space-size=5120 npm-run-all -p watch-css start-js check-version`.  

That said, it looks like this is only necessary when using the larger mock data set (available separately per licensing), not the one that's in the repo.

noranda replied:
@rdingwell Thoughts?

rdingwell replied:
@dehall does the addition cause it not to run on windows?

rdingwell replied:
@noranda the env setting crashes on windows.   Seeing how it's only needed large static data sets why dont we just remove it for now and set that env locally

dehall replied:
Yes it prevents `yarn start` from running: 
> λ yarn start
yarn run v1.3.2
$ NODE_OPTIONS=--max-old-space-size=5120 npm-run-all -p watch-css start-js check-version
'NODE_OPTIONS' is not recognized as an internal or external command,
operable program or batch file.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

rdingwell replied:
I'll go ahead an remove this and push the update

noranda replied:
I removed and pushed.

dehall commented:
Apologies I didn't have time to review this in detail before having to jet off. Overall it looks good, but I saw one small issue.  When there's nothing in the "include only" section then the rows in the "compare with" section have inconsistent behavior.  Sometimes rows in the "compare with" section were just white bars all the way across, sometimes they have green sections that are unlabeled. 

<img width="1004" alt="screen shot 2019-02-11 at 8 28 38 am" src="https://user-images.githubusercontent.com/13512036/52566296-68253c00-2dd7-11e9-9494-9c0a89e94a04.png">

Dtphelan1 commented:
[DP1] Any reason we aren't using the typical react mechanism for event listeners like the `onClick` prop and are instead manually adding the event listeners ourselves on componentDidMount? [The motif I'm referring to]( https://reactjs.org/docs/handling-events.html )

Dtphelan1 commented:
[DP2] There's 200px of bottom-margin on the `treatment-options-outcomes` class. Any reason why?

Dtphelan1 commented:
[DP3] Is there any reason why the default "Similar Patient" options are too restrictive to display any comparative treatment options?

noranda commented:
@dehall Good catch! I pushed a commit that should fix this.

@Dtphelan1 
[DP1]: The reason it's done on component mount is because the user needs to click anywhere in the body to close the poppers, meaning we can't hook an onClick to a specific element.
[DP2]: I added this margin to prevent the screen from jumping when you open the compared table popper when you only have a few rows in there.
[DP3]: The seer data we're using in the repo is a small dataset, so it's very restrictive since we've now hooked up many of the checkboxes. It's useful for testing purposes, but not representative of the large datasets we expect to use. @rdingwell and I have a larger locally stored dataset that we use for demo purposes.

Rebased.

Dtphelan1 commented:
[DP1] Makes complete sense -- my fault for not recognizing that click-anywhere-to-close behavior. Resolved!

[DP2] I see what behavior you're talking about when the popover overflows off the page. That said, the 200px of whitespace is pretty glaring both in the TDP and in the minimap. I don't think this should hold up your task, but I'd like to make a JIRA exploring writing a function that checks to see if the render is going to overflow off the page and maybe adjusts the `top` value to be a negative value, so that way the popover never causes that jumpiness.  That is, dynamically render something like this: 
![image](https://user-images.githubusercontent.com/5703736/52588419-6fb20880-2e0a-11e9-81b1-993e42b866ef.png)

[DP3] If you both have better demo data that you use locally then I feel okay, but it would be more convenient for others when testing/demoing if the default values we use now were not as restrictive, at least until we have that more expansive dataset deployed for everyone to use. I'll defer to @rdingwell or @gregquinn2001 here.

Dtphelan1 commented:
Other than these comments, this all looks good to me. I'll approve once DP3 is addressed by Rob or Greg.

noranda commented:
[DP2] Agreed the extra space is a bit noticeable, although I probably wouldn't say glaring. I would caution against your suggested fix though, as the placement of the poppers helps the user to understand which table they're editing options for. If we move the compared options up, the user may be confused as to which options they're choosing (included or compared) even with the helper text. We should probably get some ideas from Dan, but agreed that this shouldn't hold up the PR.

[DP3] @rdingwell does have a larger dataset in his transfer folder, but we don't want to add it to the repo due to its size and licensing concerns.

Dtphelan1 commented:
Re: DP3 I figured the data you mentioned wouldn't be used, the long term solution is querying a service. But until then, I defer to Rob and Greg on whether or not we should make the defaults less strict.   
  

  
Re: DP2, adjectives and connotations aside, it's a detail of the app's display that is a departure from the Mockups. Good point about the visual cues regarding the placement of the menus. Love the idea of chatting with Dan about potential alternatives. I'll make a JIRA tomorrow to implement whatever designs come out of that eventual discussion. Could you set up that meeting/communicate the nature of the problem to him? Thanks!
  

  
  

  
  
>   
> On Feb 11, 2019 at 3:16 PM,  <Noranda Brown (mailto:notifications@github.com)>  wrote:
>   
>   
>   
>
> [DP2] Agreed the extra space is a bit noticeable, although I probably wouldn't say glaring. I would caution against your suggested fix though, as the placement of the poppers helps the user to understand which table they're editing options for. If we move the compared options up, the user may be confused as to which options they're choosing (included or compared) even with the helper text. We should probably get some ideas from Dan, but agreed that this shouldn't hold up the PR.
>
>   
>
> [DP3]  @rdingwell (https://github.com/rdingwell)  does have a larger dataset in his transfer folder, but we don't want to add it to the repo due to its size and licensing concerns.
>
>   
>
> —
>  You are receiving this because you were mentioned.
>  Reply to this email directly,   view it on GitHub (https://github.com/FluxNotes/flux/pull/568#issuecomment-462478126), or   mute the thread (https://github.com/notifications/unsubscribe-auth/AFcIOO4QuZF0BN2CylowglQATXwuYcAOks5vMc-wgaJpZM4asyPT).
>
>     
>

noranda commented:
[DP2] Emailed Dan for design advice, who reminded me that we will have a legend at the bottom which should solve this issue. That task is being worked on now.

Dtphelan1 commented:
Sounds good to me!
  
  
  

  
  

  
  
>   
> On Feb 11, 2019 at 6:06 PM,  <Noranda Brown (mailto:notifications@github.com)>  wrote:
>   
>   
>   
>
> [DP2] Emailed Dan for design advice, who reminded me that we will have a legend at the bottom which should solve this issue. That task is being worked on now.
>
>   
>
> —
>  You are receiving this because you were mentioned.
>  Reply to this email directly,   view it on GitHub (https://github.com/FluxNotes/flux/pull/568#issuecomment-462531361), or   mute the thread (https://github.com/notifications/unsubscribe-auth/AFcIONR0CdffEdsXD_wYtMjqsHhp6Mhmks5vMfd3gaJpZM4asyPT).
>
>     
>

noranda commented:
Rebased. Need a second approval on this. @dehall or @Dtphelan1 ?

Dtphelan1 commented:
I'd be happy to approve once we address DP3! Pinging @rdingwell and @gregquinn2001 again for thoughts.

gregquinn2001 commented:
I think it makes sense to create a JIRA to improve the data such that the initial default settings shows results. So for DP3, I think a separate JIRA makes sense. Thoughts?

Dtphelan1 commented:
That would be a JIRA for DP3 then. Sounds good to me @gregquinn2001. I'll do that now and then approve.

Dtphelan1 commented:
[DP-4] (I'm making a JIRA so not blocking but want to post here so we're all aware.) The rows for outcomes when viewing the portrait-mode ipad viewer on chrome look very dense and are (IMO) hard to read. We should investigate the responsiveness of this application in those cases. Image below: 

![image](https://user-images.githubusercontent.com/5703736/52812499-a172db80-3065-11e9-89ac-5852aabf01eb.png)

rdingwell commented:
Apologies for not seeing this sooner. I think Greg is right with regards to improving the default data for testing but I think making the initial filters less restrictive is fine too.  I’m fine with either solution
